### ISSUE TITLE
Fix failing link tooltip Karma test

### ISSUE THREAD

**Issue by BrittanyIRL** [GoogleForCreators/web-stories-wp#6950]
These tests have been disabled: 

should display the link tooltip correctly

They should be fixed and enabled.

**Comment by BrittanyIRL** [comment]
@miina - @embarks and I are still seeing this test fail intermittently :( 

https://github.com/google/web-stories-wp/pull/6957/checks?check_run_id=2272538982

**Comment by miina** [comment]
Eh, okay. It passed 3 times in a row so thought it got fixed :/
Will disable the test in another PR again.

**Comment by samwhale** [comment]
After some investigation this is where I've gotten to:

After re-running this test in CI a few times, we finally hit this error: `Error: no node found for selector`.
![image.png](https://images.zenhubusercontent.com/5b1032a1348b7966b43421a4/14eea22f-cdf9-4feb-8428-e084d7fc84cf)

This bug can generally be fixed with `page.waitForSelector` but it looks like the tests don't have access to puppeteer's `page` object, right?

Initially it was not reproducible locally, but that was because we were running it in a browser. This test fails consistently when run locally in `headless` mode.

When run locally in headless mode, we see a slightly different error: `Error: Node is either not visible or not an HTMLElement` which suggests that the element is not in puppeteers viewport. The [puppeteer docs](https://github.com/puppeteer/puppeteer/blob/v8.0.0/docs/api.md#elementhandleclickoptions) state that the `click` method should scroll the element into view though, and that works when testing in the browser. I'm hoping that both errors are coming from the same problem, since they're similar errors.

It seems to me that the `fixture.events.click` method is not scrolling the `link` input into view when in headless mode before we try to click. I've found [some threads](https://github.com/puppeteer/puppeteer/issues/2977#issuecomment-412807613) online that suggest using `page.evaulate` to ensure that an element is scrolled into view. However, I've tried using `element.scrollIntoView` and have had no luck.

@miina @swissspidy I want to try using the `page.evaluate` method to try to fix this text but I don't know where puppeteer's `page` object is being kept in our custom puppeteer setup. Do y'all have any ideas about how we could get access to the `page` object?

**Comment by swissspidy** [comment]
> This bug can generally be fixed with `page.waitForSelector` but it looks like the tests don't have access to puppeteer's `page` object, right?

Not at the moment. We use  Testing Library's `waitFor` function most of the time for that, e.g. `await waitFor(() => fixture.screen.getByRole(...))`

> When run locally in headless mode, we see a slightly different error: `Error: Node is either not visible or not an HTMLElement` which suggests that the element is not in puppeteers viewport. The [puppeteer docs](https://github.com/puppeteer/puppeteer/blob/v8.0.0/docs/api.md?rgh-link-date=2021-04-09T15%3A28%3A27Z#elementhandleclickoptions) state that the `click` method should scroll the element into view though, and that works when testing in the browser. I'm hoping that both errors are coming from the same problem, since they're similar errors.

Puppeteer does indeed scroll. And perhaps the issue in this case is _not_ the scrolling.

> It seems to me that the `fixture.events.click` method is not scrolling the `link` input into view when in headless mode before we try to click. I've found [some threads](https://github.com/puppeteer/puppeteer/issues/2977#issuecomment-412807613) online that suggest using `page.evaluate` to ensure that an element is scrolled into view. However, I've tried using `element.scrollIntoView` and have had no luck.

[This comment](https://github.com/puppeteer/puppeteer/issues/2977#issuecomment-609415987) from that thread you linked to seems very relevant:

> > _So, `.click()` call is failing likely because the pointed element actually has the 0x0 rectangle, which often happens when element is simply a wrapper._

This seems worth testing.

> I want to try using the `page.evaluate` method to try to fix this text but I don't know where puppeteer's `page` object is being kept in our custom puppeteer setup. Do y'all have any ideas about how we could get access to the `page` object?

I'm not sure you really need that.

`page.evaluate()` simply evals a function in the page context (i.e. in the dev tools browser console). But in the Karma tests you're already in the page context and can access elements using the Testing Library utils and the fixture.

In other words, you don't need to do `page.evaluate(()=> ...)` when you can just do `document.querySelector(...)` or `fixture.screen.getBy..` in tests.

FWIW, `karma-puppeteer-launcher` and `karma-puppeteer-client` are the packages exposing Puppeteer functions to the test suite (then consumed by `karma-fixture` / `fixture.events`

**Comment by samwhale** [comment]
Thanks for the well written reply @swissspidy. It really helped narrow down what I was doing wrong.

I was running `npm run test:karma:edit-story -- --headless` but it was failing because the viewport size was not being set. I assumed that the config in `jest-puppeteer.config.cjs` was setting the viewport, but I was wrong. Running the test with `npm run test:karma-edit-story -- --headless --viewport=1600:1000` made the tests pass ✅ 

Ended up adding some `waitFor`s to make sure the elements were rendered before trying to click them.
